examples/async2-passthru.rs

```diff
@@ -0,0 +1,447 @@
+        let driver = self.driver.clone();
+
+        // first try to read packet
+        if unsafe { driver.read_packet(packet) }.is_ok() {
+            return Ok(());
+        }
+
+        let result = self.wait_for_packet().await; // wait for packet event
+        match result {
+            Ok(_) => {
+                if unsafe { driver.read_packet(packet) }.ok().is_some() {
+                    Ok(())
+                } else {
+                    Err(unsafe { GetLastError() }.into())
+                }
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+// Implementing the Drop trait for the NdisapiAdapter struct.
+impl Drop for NdisapiAdapter {
+    /// The drop method will be called automatically when the NdisapiAdapter object goes out of scope.
+    fn drop(&mut self) {
+        // Setting the operating mode for the specified adapter to default.
+        _ = self
+            .driver
+            .set_adapter_mode(self.adapter_handle, FilterFlags::from_bits_truncate(0));
+
+        // Setting the packet event for the specified adapter to NULL.
+        _ = self.driver.set_packet_event(self.adapter_handle, HANDLE(0));
+    }
+}
+
+/// A future that resolves when a Win32 event is signaled.
+struct Win32EventFuture {
+    #[allow(dead_code)]
+    notif: Win32EventNotification, // The Win32 event notification object.
+    waker: Arc<AtomicWaker>,       // An atomic waker for waking the future.
+    ready: Arc<AtomicBool>,        // An atomic boolean indicating whether the event is ready.
+}
+
+impl Win32EventFuture {
+    /// Create a new `Win32EventFuture` instance with the specified event handle.
+    fn new(event_handle: HANDLE) -> Result<Self> {
+        let waker = Arc::new(AtomicWaker::new());
+        let ready = Arc::new(AtomicBool::new(false));
+
+        Ok(Self {
+            waker: waker.clone(),
+            ready: ready.clone(),
+            notif: Win32EventNotification::new(
+                event_handle,
+                Box::new(move |_| {
+                    println!("=========================================================================================================================================");
+                    ready.store(true, Ordering::SeqCst);
+                    waker.wake();
+                    unsafe { ResetEvent(event_handle) };
+                }),
+            )?,
+        })
+    }
+
+    /// Poll for the packet event.
+    pub fn poll_packet_event(&mut self, cx: &mut Context) -> Poll<Result<()>> {
+        if self.ready.swap(false, Ordering::Relaxed) {
+            Poll::Ready(Ok(()))
+        } else {
+            self.waker.register(cx.waker());
+            Poll::Pending
+        }
+    }
+}
+
+impl Future for Win32EventFuture {
+    type Output = Result<()>;
+
+    /// Poll the future to check if the event is ready.
+    fn poll(self: std::pin::Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        Pin::into_inner(self).poll_packet_event(cx)
+    }
+}
```

Comment:
Futures typically only resolve once. Consider not implementing `Future` and only expose the the `poll` function. You can always create an ad-hoc `Future` with `futures::future::poll_fn`.

Reply:
I think implementing [`Stream`](https://docs.rs/futures/latest/futures/stream/trait.Stream.html) would be the best way to go here.